packages/eslint-plugin-query/src/rules/stable-query-client/stable-query-client.rule.ts

```diff
@@ -0,0 +1,88 @@
+import { AST_NODE_TYPES } from '@typescript-eslint/utils'
+import { ASTUtils } from '../../utils/ast-utils'
+import { createRule } from '../../utils/create-rule'
+import type { TSESLint } from '@typescript-eslint/utils'
+
+export const name = 'stable-query-client'
+
+export const rule = createRule({
+  name,
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Makes sure that QueryClient is stable',
+      recommended: 'error',
+    },
+    messages: {
+      unstable: [
+        'QueryClient is not stable. It should be either extracted from the component or wrapped in React.useState.',
+        'See https://tkdodo.eu/blog/react-query-fa-qs#2-the-queryclient-is-not-stable',
+      ].join('\n'),
+      fixTo: 'Fix to {{result}}',
+    },
+    hasSuggestions: true,
+    fixable: 'code',
+    schema: [],
+  },
+  defaultOptions: [],
+
+  create(context, _, helpers) {
+    return {
+      NewExpression(node) {
+        if (
+          node.callee.type !== AST_NODE_TYPES.Identifier ||
+          node.callee.name !== 'QueryClient' ||
+          !helpers.isSpecificTanstackQueryImport(
+            node.callee,
+            '@tanstack/react-query',
+          )
+        ) {
+          return
+        }
+
+        const fnAncestor = ASTUtils.getFunctionAncestor(context)
+
+        if (!ASTUtils.isValidReactComponentOrHookName(fnAncestor?.id)) {
+          return
+        }
+
+        const declarator = ASTUtils.getClosestVariableDeclarator(node)
+
+        if (
+          declarator?.init?.type === AST_NODE_TYPES.CallExpression &&
+          ['useState', 'React.useState'].includes(
```

Comment:
I guess it would also work inside `useRef`, but that's hard to find I guess?

```
const ref = useRef()
if (!ref.current) {
  ref.current = new QueryClient()
}
```

Reply:
Oh, It's doable of course, but is this a real use case?

Reply:
interestingly, there are lots of users who like to use refs for this. There's also a popular `useConstant` package that does it like that:

https://github.com/Andarist/use-constant/blob/bdd74e6348c14b7947c278fa5a5abdc03a635ac2/src/index.ts#L5-L13

come to think about it, maybe we should only error if `new QueryClient()` is called directly in the render function of a component, without being wrapped in anything? Because:

```
const client = useConstant(() => new QueryClient())
```

should also work.

Reply:
Sounds logical. Even this is considered valid:

```
const client = React.useMemo(() => new QueryClient(), [])
```

I'll update the PR accordingly.

Reply:
thanks. another thing I just realised: `new QueryClient()` is actually okay, and recommended, inside server components (async components). From our docs:

```
export default async function PostsPage() {
  const queryClient = new QueryClient()

  await queryClient.prefetchQuery({
    queryKey: ['posts'],
    queryFn: getPosts,
  })
```

this is fine because server components never re-render, and if we'd put the QueryClient outside of the component, it would be shared between users and requests. Can we account for that?

Reply:
Sure, that would make sense. I have updated the PR accordingly.